File: ai/6/6.py
import heapq

def main():
    tt = int(input())
    for _ in range(tt):
        n, e = map(int, input().split())
        edges = [[] for _ in range(n)]
        for _ in range(e):
            u, v = map(int, input().split())
            edges[u].append(v)
            edges[v].append(u)
        h = list(map(int, input().split()))
        start, end = map(int, input().split())
        pq = []
        heapq.heappush(pq, (h[start], start))
        
        parent = [-1] * n
        visited = [False] * n
        visited[start] = True
        
        traversal = []
        found = False
        
        while pq:
            curr_h, curr = heapq.heappop(pq)
            traversal.append(curr)
            if curr == end:
                found = True
                break
            for neighbor in edges[curr]:
                if not visited[neighbor]:
                    heapq.heappush(pq, (h[neighbor], neighbor))
                    visited[neighbor] = True
                    parent[neighbor] = curr
        
        print("Traversal:", " ".join(map(str, traversal)))
        if not found:
            print("Path: No path found")
        else:
            path = []
            cur = end
            cost = 0
            while cur != -1:
                cost += h[cur]
                path.append(cur)
                cur = parent[cur]
            path.reverse()
            print("Path:", " ".join(map(str, path)))
            print("Total cost: ", cost)
            
if __name__ == "__main__":
    main()

'''
OUTPUT:
1
7 10
1 2
1 3
2 3
2 4
3 5
3 7
4 5
4 6
5 6
6 7
1 2 2 6 3 6 4
6 1
Traversal: 6 2 0 1 
Path: 6 2 1 
Total cost: 8
'''

File: ai/6/6.cpp
#include <bits/stdc++.h>
using namespace std;

signed main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int tt;
    cin >> tt;
    while (tt--) {
        int n, e;
        cin >> n >> e;
        vector<vector<int>> edges(n);
        
        for (int i = 0; i < e; i++) {
            int u, v;
            cin >> u >> v;
            edges[u].push_back(v);
            edges[v].push_back(u);
        }
        
        vector<int> h(n);
        for (int i = 0; i < n; i++) {
            cin >> h[i];
        }
    
        int start, end;
        cin >> start >> end;
        
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        pq.push({h[start], start});
        
        vector<int> parent(n, -1);
        vector<bool> visited(n, false);
        visited[start] = true;
        
        vector<int> traversal;
        bool found = false;
        
        while (!pq.empty()) {
            auto [curr_h, curr] = pq.top();
            pq.pop();
            traversal.push_back(curr);
            
            if (curr == end) {
                found = true;
                break;
            }
            
            for (int neighbor : edges[curr]) {
                if (!visited[neighbor]) {
                    pq.push({h[neighbor], neighbor});
                    visited[neighbor] = true;
                    parent[neighbor] = curr;
                }
            }
        }
        
        cout << "Traversal: ";
        for (int node : traversal) {
            cout << node << " ";
        }
        cout << endl;

        if (!found) {
            cout << "Path: No path found" << endl;
        } else {
            vector<int> path;
            int cur = end, cost = 0;
            while (cur != -1) {
                cost += h[cur];
                path.push_back(cur);
                cur = parent[cur];
            }
            reverse(path.begin(), path.end());
            
            cout << "Path: ";
            for (int node : path) {
                cout << node << " ";
            }
            cout << endl;
            cout << "Total cost: " << cost << endl;
        }
    }
    
    return 0;
}
